compiler/rustc_target/src/spec/mod.rs

```diff
@@ -3709,6 +3709,18 @@ impl TargetTriple {
             }
         }
     }
+
+    /// Helper function to check if the given `triple` ends with `".json"` __case-insensitive__.
+    /// Does not check if the path exists.
+    pub fn has_json_ext<S: AsRef<str>>(triple: S) -> bool {
+        const EXT: &'static str = ".json";
+        triple.as_ref().ends_with(EXT) || {
+            let s = triple.as_ref();
+            s.get(s.len().wrapping_sub(EXT.len())..)
+                .map(|ext| ext.eq_ignore_ascii_case(EXT))
+                .unwrap_or_default()
+        }
+    }
```

Comment:
You could just let `Path` do the work here, `Path::new(triple).extension().is_some_and(|ext| ext.eq_ignore_ascii_case(EXT))`.

But maybe better just to name this `has_ext_ignore_case` and take the desired extension as an argument? So it could be reused with other files, (probably doesn't need to be in `TargetTriple` either since there's nothing triple-specific).

Reply:
Thank you for reviewing.

> You could just let Path do the work here, Path::new(triple).extension().is_some_and(|ext| ext.eq_ignore_ascii_case(EXT)).

I tried to optimise, so in case of ext is `"json"` it costs exactly same as before my patch. 
But now I see, double cost if there is no `".json"` 🤔 You totally right, so for your suggestion in approximately middle case cost will be a little bit higher then before the patch, but definitely not double.

Reply:
Done.


So we have this: (actual for any input)
 - `file_name`: `components().next_back().and_then(/* closure with match */)`
 - `extension`: `file_name` + `map(rsplit_file_at_dot).and_then(|(opta, optb)| opta.and(optb))`

Imho, in some future could be good to optimise it like following:
 - take slice of bytes by the input
 - take slice of last {len of ext} chars
   - if can't => false
 - `eq_ignore_ascii_case` for that slice and ext
 It should be slightly better in cost.

However, the price is so low that with your permission, I might look into it later with a new PR.